src/frontend/apps/standalone_site/src/components/SortableTable/index.tsx

```diff
@@ -0,0 +1,266 @@
+import { Box, CheckBox, Menu, Pagination, Stack, Text } from 'grommet';
+import { Maybe } from 'lib-common';
+import { Spinner } from 'lib-components';
+import { Fragment, ReactNode, useEffect, useState } from 'react';
+
+interface SortByBase {
+  label: string;
+}
+
+interface SortableTableBase<ItemType> {
+  title: React.ReactNode;
+  loading?: boolean;
+  items: (() => Promise<ItemType[]>) | ItemType[];
+  children: (item: ItemType) => ReactNode;
+}
+
+interface UnSortable {
+  sortable?: false;
+}
+interface Sortable<ItemType, SortByType extends SortByBase = SortByBase> {
+  sortable: true;
+  currentSort: SortByType;
+  sortBy: SortByType[];
+  onSortChange: (newSort: SortByType) => ItemType[] | Promise<ItemType[]>;
+}
+
+interface UnSelectable {
+  selectable?: false;
+}
+interface Selectable<ItemType> {
+  selectable: true;
+  onSelectionChange: (newSelection: ItemType[]) => void;
+}
+
+interface UnPaginable {
+  paginable?: false;
+}
+interface Paginable<ItemType> {
+  paginable: true;
+  currentPage?: number;
+  numberOfItems: number;
+  pageSize?: number;
+  onPageChange: (newPage: number) => ItemType[] | Promise<ItemType[]>;
+}
+
+type SortableTableProps<
+  ItemType,
+  SortByType extends SortByBase = SortByBase,
+> = SortableTableBase<ItemType> &
+  (UnSortable | Sortable<ItemType, SortByType>) &
+  (Selectable<ItemType> | UnSelectable) &
+  (Paginable<ItemType> | UnPaginable);
+
+export const SortableTable = <
+  ItemType,
+  SortByType extends SortByBase = SortByBase,
+>({
+  title,
+  loading: initialLoading,
+  items: initialItems,
+  children,
+  ...props
+}: SortableTableProps<ItemType, SortByType>) => {
+  const [isLoading, setIsLoading] = useState(true);
+  const [items, setItems] = useState<ItemType[]>([]);
+  const [previousSelection, setPreviousSelection] = useState<ItemType[]>([]);
+  const [currentSelection, setCurrentSelection] = useState<ItemType[]>([]);
+  const [targetSort, setTargetSort] = useState<Maybe<SortByType>>(undefined);
+  const [asyncLoader, setAsyncLoader] =
+    useState<Maybe<() => Promise<ItemType[]>>>(undefined);
+
+  let selectionCallback: Maybe<Selectable<ItemType>['onSelectionChange']> =
+    undefined;
+  if (props.selectable) {
+    selectionCallback = props.onSelectionChange;
+  }
+
+  let sortCallback: Maybe<Sortable<ItemType, SortByType>['onSortChange']> =
+    undefined;
+  let SortMenu: Maybe<JSX.Element>;
```

Comment:
Because a hook can not be conditional (so I can not check first `props.sortable`) and else I will have to put props in the conditions and that will also cause it to be redraw ofter

Reply:
As it is now, the Menu component could potentially rerender many times if sortable is defined.

```suggestion
  let sortItems: SortByType[] | undefined = undefined;
  let sortCurrentItem: SortByType | undefined = undefined;

  if (props.sortable) {
    sortCallback = props.onSortChange;
    sortItems = props.sortBy;
    sortCurrentItem = props.currentSort;
  }

  const SortMenu = useMemo(() => {
    if (!sortCallback || !sortItems || !sortCurrentItem) {
      return;
    }

    const availableItems = sortItems.filter((item) => item !== sortCurrentItem);

    return (
      <Menu
        a11yTitle="Sort item in the table"
        margin={{ left: 'auto', vertical: 'auto' }}
        label={sortCurrentItem.label}
        items={availableItems.map((item) => ({
          label: item.label,
          onClick: () => {
            setTargetSort(item);
          },
        }))}
      />
    );
  }, [sortCallback, sortItems, sortCurrentItem]);
```

Something like that, although Im not sure about this condition : 
```
if (!sortCallback || !sortItems || !sortCurrentItem) {
      return;
    }
```

Reply:
Yes it will be recreated every render of the component if sortable is defined but is that even an issue ?
Unless we face render issue I think this solution kinda hurts readability which is already not great

Reply:
You must aim in general to make you code performant no? What if Menu get greedy in the future? 
It is just a small useMemo caching a component or not moreover, I don't see any problem of readability. 

If you like it like like before no problem.

Reply:
No I disagree, of course you have to avoid creating bottleneck but if performances come at a cost in readability and therefore maintainability it is uncessecary until required.